[14기 이승효]  step1 돔 조작과 이벤트 핸들링으로 메뉴 관리하기

bingwer opened the pull request:
- [x] 에스프레소 메뉴에 새로운 메뉴를 확인 버튼 또는 엔터키 입력으로 추가한다.
  - [x] 메뉴가 추가되고 나면, input은 빈 값으로 초기화한다.
  - [x] 사용자 입력값이 빈 값이라면 추가되지 않는다.
- [x] 메뉴의 수정 버튼을 눌러 메뉴 이름 수정할 수 있다.
  - [x] 메뉴 수정시 브라우저에서 제공하는 `prompt` 인터페이스를 활용한다.
- [x] 메뉴 삭제 버튼을 이용하여 메뉴 삭제할 수 있다.
  - [x] 메뉴 삭제시 브라우저에서 제공하는 `confirm` 인터페이스를 활용한다.
- [x] 총 메뉴 갯수를 count하여 상단에 보여준다.
- 추가되는 메뉴의 아래 마크업은 `<ul id="espresso-menu-list" class="mt-3 pl-0"></ul>` 안에 삽입해야 한다.
---------------------------------
아쉬운 점 : 
1. 컴포넌트를 좀 더 추상화 하여, 메뉴 form 부분과 list를 분할하고 싶었는데 그러지 못해 아쉽다.
2. render() 관련 부분을 각 컴포넌트마다 따로 구현하도록 하였는데, 이것도 추상화가 가능할 것 같아 시도해보았으나 해결하지 못해 아쉽다.

bingwer committed:
Docs: 팀원 이름 추가하기

춘식이는 개발개발해 팀원 이름 추가

bingwer committed:
feat) 1주차 미션(step1) 기능구현
- 에스프레소 메뉴에 새로운 메뉴를 확인 버튼 또는 엔터키 입력으로 추가한다.
- 메뉴가 추가되고 나면, input은 빈 값으로 초기화한다.
- 사용자 입력값이 빈 값이라면 추가되지 않는다.
- 메뉴의 수정 버튼을 눌러 메뉴 이름 수정할 수 있다.
- 메뉴 수정시 브라우저에서 제공하는 `prompt` 인터페이스를 활용한다.
- 메뉴 삭제 버튼을 이용하여 메뉴 삭제할 수 있다.
- 메뉴 삭제시 브라우저에서 제공하는 `confirm` 인터페이스를 활용한다.
- 총 메뉴 갯수를 count하여 상단에 보여준다.
- 추가되는 메뉴의 아래 마크업은 `<ul id="espresso-menu-list" class="mt-3 pl-0"></ul>` 안에 삽입해야 한다.

bingwer committed:
refactor) prompt 함수 refactor

bingwer committed:
refactor) 파일명 변경, 전역 store 삭제

bingwer committed:
refactor) state를 직접 조작하는 부분 코드 수정

zwonlala commented on src/js/constant/index.js:
```diff
@@ -0,0 +1,5 @@
+export const alertText = Object.freeze({
```
오 alert() 에 사용되는 문자열들을 따로 뽑아 관리하는 구조로 구현하셨군요 👍 

뭔가 상수처럼 사용되는 느낌인데 "alertText"도 "ALERT_TEXT"와 같이 네이밍 하는 건 어떤가요? 🤔

bingwer replied:
객체 내 key값은 네이밍 잘해두고 정작 객체명자체를 네이밍을 잘못했네요 😂 감사합니다!

zwonlala commented on src/js/components/Component.js:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    this.state = {};
+    this.container = $container;
```
변수명만으로는 아래 

- baseTemplate
- renderTemplate
의 차이에 대해 잘 모르겠는데 혹시 설명해주실 수 있나요?

bingwer replied:
- baseTemplate : 수정되기전의 HTML Template입니다.
- renderTemplate: updateTemplate 함수를 통해 메뉴이름이 변경된 template입니다.

저는 renderTemplate에 각각 event를 붙여주는 방식으로 렌더링하였고, 그 후에 사용된 renderTemplate를 다시 baseTemplate로 초기화 해주는 방식으로 구현했어요!

zwonlala replied:
@bingwer 혹시 설명해주신 부분과 관련된 개념이라던가, 비슷한 구조를 가지는 게 있나요?
관련 된 내용에 대한 글 같은거 공유해주시면 감사하겠습니다! 🙏

bingwer replied:
@zwonlala 관련된 개념이라기보다는.. Template을 이용하여 비즈니스로직과 프레젠테이션 부분을 분리해주는 라이브러리중에 Handlebars라는 라이브러리가 있는데요. 해당 라이브러리를 이용하는것과 최대한 비슷하게 구현을 해보았어요..!!

[Handlebars](https://handlebarsjs.com/)

zwonlala commented on src/js/components/Menu.js:
```diff
@@ -0,0 +1,110 @@
+
+export default class Menu extends Component {
+  constructor(containerId) {
+    const menuListTemplate = `<li class="menu-list-item d-flex items-center py-2">
```
뭔가 생성자 내 변수보다는
class의 static 필드로 정의되면 더 좋을 것 같다는 생각이 드는데 승효님께서는 어떻게 생각하시나요?

bingwer replied:
@zwonlala  static 필드는 보통 클래스의 인스턴스 생성 없이 사용가능한 util함수 등을 만들때 사용되는 것으로 알고있는데요.
저는 해당 부분은 단순히 생성자 초기화시에만 필요할것같아서 단순 변수로 선언했어요!
혹시 static으로 선언했을때 얻게되는 또다른 장점이 있을까요?

zwonlala replied:
static 필드로 선언시 얻게되는 장점이라기 보단,

해당 템플릿이 클래스의 인스턴스 별로 서로 다른 값을 가질 수 있는게 아니라 
- 해당 클래스의 인스턴스이면 동일한 값을 사용하는? 
- 인스턴스에 종속된 값이 아닌, 해당 클래스에 종속된 값이라?

static으로 선언하는게 자연스럽다고 생각했어요.

zwonlala commented on src/js/components/Menu.js:
```diff
@@ -0,0 +1,110 @@
+  addMenu() {
+    const $menuInput = document.getElementById('espresso-menu-name');
+    const newMenu = $menuInput.value;
+    if (newMenu.replaceAll(' ', '') === '') {
```
옹 String.prototype.trim() 메소드가 아닌 ".replaceAll(' ', '')" 을 사용하신 이유가 있나요?

bingwer replied:
저는 메뉴이름이 중복되는걸 피하고자 해당 방법을 사용했어요.
아이스라떼 <-> 아이스 라떼 
이런식의 공백으로 인한 중복을 피하고 싶었습니다. 
trim()메소드는 문자열의 앞뒤 공백만 없애주기때문에 replaceAll 메소드를 이용했어요!

zwonlala replied:
옹 이런 뜻이 있었군요. 🤔 

이번 Step1에 중복을 제거하는 처리가 포함되어있진 않았어서 깊게 고민해보진 않았지만 

저는 사용자가 중간에 "아이스 라떼" 이렇게 작성을 한 경우라면 
"아이스라떼" 와는 다르게 처리되어야 한다고? 생각했었습니다.

사용자가 '의도'를 가지고 공백을 추가했다고 생각이 들어서요~

zwonlala commented on src/js/components/Component.js:
```diff
@@ -0,0 +1,70 @@
+    this.container.innerHTML = '';
+  }
+
+  makeConfirmAlert(msg) {
```
개인적으로 
- confirm
- prompt
관련 로직을 Component 클래스 내에 존재하는게 살짝 어색하다고 생각이 드는데 승효님께서는 어떻게 생각하시나요?

bingwer replied:
컴포넌트들중에서는 저 함수들을 사용하지 않는 컴포넌트도 있을 수 있다는 생각을 하지 못했네요.
util함수로 빼서 관리하면 좀더 효과적일것 같습니다. 좋은 의견 감사합니다!

hey210056 commented on src/js/components/Component.js:
```diff
@@ -0,0 +1,70 @@
+  getHTMLElement(template) {
+    const $wrapper = document.createElement('div');
+    $wrapper.innerHTML = template;
+    const $targetElement = $wrapper.firstChild.cloneNode(true);
```
cloneNode 메서드에 대해 배워갑니다👍

hey210056 commented on src/js/components/Component.js:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  updateTemplate(key, value) {
+    this.renderTemplate = this.renderTemplate.replace(key, value);
```
key값과 value로 값을 저장해서 replace로 render하는 방법 좋은 거 같아요!

pers0n4 replied:
`renderTemplate` 속성은 template literal 형태의 string으로 사용하려는 목적이라 추론되는데, `updateTemplate`메서드에서 사용하는 `key`와 `value`라는 네이밍이 적절한가? 라는 의문이 들어요.
그리고 만약 `updateTemplate("button", "newButton")`의 형태로 호출하게 된다면 element인 `button`을 대체하는 문제가 발생하진 않을까요?

bingwer replied:
@pers0n4 저는 템플릿에서 업데이트 할 부분을 key라고 생각하고 해당값을 업데이트 한다는 생각으로 파라미터명을 key와 value로 지정을 했었는데요. 동영님께서 생각하시는 네이밍은 무엇인지 궁금해요!
또한 아래 말씀해주신 부분은 저도 생각하지 못한 부분이이였는데 정말 감사합니다🥰
key값을 메소드 호출에서가 아닌 메소드 내에서 붙이는 방법으로 존재하지 않는 key값일 경우 변경되지 않게 처리하도록 리팩토링 해봐야겠네요!

pers0n4 commented on src/js/components/Component.js:
```diff
@@ -0,0 +1,70 @@
+    this.render();
+  }
+
+  addEvent($element, eventName, func) {
+    $element.addEventListener(eventName, func);
+  }
```
이렇게 선언해두면 상속받는 클래스에서 `this.addEvent($element, eventName, func)`의 형태로 리스너를 등록하게 될 텐데, `$element.addEventListener(eventName, func)`로 직접 호출하는 것 대비 어떤 이점을 얻을 수 있을까요?

bingwer replied:
해당부분은 추상화과정에서 원래 클래스 내 renderElement에 event를 부착하려고 만들었던 메소드인데요.
결론적으로는 불필요한 메소드가 되어버렸군요.. 😓

bingwer committed:
refactor) review를 바탕으로한 리팩토링
1. 알림 함수들 utils함수로 관리
2. Html text 변경 : textContent로 변경
3. 메뉴 수정 및 삭제 이벤트 위임 방식으로 변경

bingwer commented:
@imakerjun 메이커준님 merge 부탁드립니다!

bingwer committed:
feat) 메뉴 카테고리 선택을 위한 컴포넌트 추가

bingwer committed:
Revert "feat) 메뉴 카테고리 선택을 위한 컴포넌트 추가"

This reverts commit 8441928e82c9deeeab02115b9d75fc496a8f24a7.
